src/java/main/org/apache/zookeeper/server/NettyServerCnxnFactory.java

```diff
@@ -109,6 +110,20 @@ public void channelConnected(ChannelHandlerContext ctx,
                     zkServer, NettyServerCnxnFactory.this);
             ctx.setAttachment(cnxn);
 
+            InetAddress addr = ((InetSocketAddress)cnxn.channel.getRemoteAddress()).getAddress();
+            Set<NettyServerCnxn> s = ipMap.get(addr);
+            if (s != null) {
+                synchronized (s) {
```

Comment:
If we change `s` to be a synchronized set I think we can get rid of this synchronized block.